### PR TITLE
tags support

### LINKED ISSUE #69: elbv3 listener (maybe all resources) tags value error
When listener with tags, can't fetch or delete it

openstack --os-cloud octavia-nl loadbalancer listener show 4efca02a-f43b-463a-bf39-9c56a6a41ec8 --debug
RESP BODY: {"faultcode": "Server", "faultstring": "Provider 'elbv3' reports error: Invalid input for field/attribute tags. Value: '[{'key': 'key15', 'value': 'value15'}]'. Value should be string", "debuginfo": null}

openstack --os-cloud octavia-nl loadbalancer listener delete 8a0b58cf-87ee-4bcb-bc0e-5dc5afa2659a --debug
RESP BODY: {"faultcode": "Server", "faultstring": "Provider 'elbv3' reports error: Invalid input for field/attribute tags. Value: '[{'key': 'key15', 'value': 'value15'}]'. Value should be string", "debuginfo": null}
GET call to https://octavia-nl.eco.tsi-dev.otc-service.com/v2.0/lbaas/listeners used request id req-16ef2aa1-62ed-4887-863a-b54293e1b177
Request returned failure status: 500
Unrecognized schema in response body. (HTTP 500) (Request-ID: req-16ef2aa1-62ed-4887-863a-b54293e1b177)

### PR DESCRIPTION
**Pull request by anton-sidelnikov**
Closes #69

### CODE REVIEW

File: octavia_proxy/api/drivers/elbv3/driver.py [hunk 1]
```diff
@@ -39,18 +39,24 @@ def get_supported_availability_zone_metadata(self):
...
-        return lb
+            for tag in otc_tags:
+                tags.append('%s=%s' % (tag['key'], tag['value']))
+            resource.tags = tags
+        return resource
+
+    def _resource_tags(self, tags):
+        result = []
+        for tag in tags:
+            tag = tag.split('=')
```
Comment (gtema) [inline octavia_proxy/api/drivers/elbv3/driver.py:54]: User might really pass `white` as a tag, thus split fails.

File: octavia_proxy/api/drivers/elbv3/driver.py [hunk 2]
```diff
@@ -39,18 +39,24 @@ def get_supported_availability_zone_metadata(self):
...
+    def _normalize_tags(self, resource):
+        otc_tags = resource.tags
         if otc_tags:
             tags = []
-            for k, v in otc_tags:
-                tags.append('%s=%s' % (k, v))
-            lb.tags = tags
-        return lb
+            for tag in otc_tags:
+                tags.append('%s=%s' % (tag['key'], tag['value']))
```
Comment (gtema) [inline octavia_proxy/api/drivers/elbv3/driver.py:47]: we were discussing a bit to drop `=`sign when value is empty (tag is like a plain label)

File: octavia_proxy/api/drivers/elbv3/driver.py [hunk 3]
```diff
@@ -39,18 +39,24 @@ def get_supported_availability_zone_metadata(self):
         return {"eu-nl-01": "The compute availability zone to use for "
                             "this loadbalancer."}
 
-    def _normalize_lb(self, lb):
```
Comment (gtema) [inline octavia_proxy/api/drivers/elbv3/driver.py:42]: I do not think it makes sense dropping this "normalize" function, cause we are going to have another normalization on flavors and az.

**Review by gtema — CHANGES_REQUESTED** [review]
besides comments inline, would it be also possible to get some tests for tags?

File: octavia_proxy/api/drivers/elbv3/driver.py (hunk 1 above)
Comment (anton-sidelnikov) [reply octavia_proxy/api/drivers/elbv3/driver.py:54]: https://prnt.sc/1ttnkpt

Comment (anton-sidelnikov) [reply octavia_proxy/api/drivers/elbv3/driver.py:54]: i dont get, why it fails should fail

Comment (gtema) [reply octavia_proxy/api/drivers/elbv3/driver.py:54]:
```
>>> s = "white".split("=")
>>> s[1]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IndexError: list index out of range
```

**Comment by kucerakk** [comment]
> besides comments inline, would it be also possible to get some tests for tags?

added two testcases (one for LB and one for Listener) with tags in https://github.com/opentelekomcloud-infra/octavia-proxy/pull/70, but it expects that we have full support for tags including creation of resource

File: octavia_proxy/api/drivers/elbv3/driver.py (hunk 2 above)
Comment (anton-sidelnikov) [reply octavia_proxy/api/drivers/elbv3/driver.py]: done

File: octavia_proxy/api/drivers/elbv3/driver.py (hunk 3 above)
Comment (anton-sidelnikov) [reply octavia_proxy/api/drivers/elbv3/driver.py:42]: ok, revert change

File: octavia_proxy/api/drivers/elbv3/driver.py (hunk 1 above)
Comment (anton-sidelnikov) [reply octavia_proxy/api/drivers/elbv3/driver.py]: done

**Comment by anton-sidelnikov** [comment]
> > besides comments inline, would it be also possible to get some tests for tags?
> 
> added two testcases (one for LB and one for Listener) with tags in #70, but it expects that we have full support for tags including creation of resource

Tags fully supported for lb and listener .added unit tests.

File: octavia_proxy/api/drivers/elbv3/driver.py [hunk 4]
```diff
@@ -98,13 +113,15 @@ def loadbalancer_create(self, session, loadbalancer):
         LOG.debug('Creating loadbalancer %s' % loadbalancer.to_dict())
 
         lb_attrs = loadbalancer.to_dict()
+        if lb_attrs['tags']:
```
Comment (kucerakk) [inline octavia_proxy/api/drivers/elbv3/driver.py:116]: this fails with KeyError if the lb_attrs has no 'tags' key, rather use ``` if 'tags' in lb_attrs ```

File: octavia_proxy/api/drivers/elbv3/driver.py [hunk 5]
```diff
@@ -184,11 +204,14 @@ def listener_create(self, session, listener):
         if 'timeout_member_connect' in lattrs:
             lattrs['keepalive_timeout'] = \
                 lattrs.pop('timeout_member_connect')
+        if lattrs['tags']:
```
Comment (kucerakk) [inline octavia_proxy/api/drivers/elbv3/driver.py:207]: same here... this fails with KeyError if the lattrs has no 'tags' key, rather use ``` if 'tags' in lattrs ```

**Comment by kucerakk** [comment]
``` def create_member() ``` method in elbv3 driver should be extended...we discussed that if the subnet_id is not provided, than it is taken from the ELB...something like (similar to elbv2 driver)

``` 
if 'subnet_id' in attrs:
            attrs['subnet_cidr_id'] = attrs.pop('subnet_id')
else:
            lb_id = session.vlb.get_pool(pool_id)['loadbalancers'][0]['id']
            attrs['subnet_cidr_id'] = session.vlb.get_load_balancer(
                lb_id)['vip_subnet_id']				
```

**Comment by kucerakk** [comment]
with ELBv3 the tags testcases written in https://github.com/opentelekomcloud-infra/octavia-proxy/pull/70 passes (obviously after updating the code according my previous comments)

**Comment by anton-sidelnikov** [comment]
> `def create_member()` method in elbv3 driver should be extended...we discussed that if the subnet_id is not provided, then it is taken from the ELB...something like (similar to elbv2 driver)
> 
> ```
> if 'subnet_id' in attrs:
>             attrs['subnet_cidr_id'] = attrs.pop('subnet_id')
> else:
>             lb_id = session.vlb.get_pool(pool_id)['loadbalancers'][0]['id']
>             attrs['subnet_cidr_id'] = session.vlb.get_load_balancer(
>                 lb_id)['vip_subnet_id']				
> ```
 Yep, done.

**Review by lego963 — DISMISSED** [review]

**Review by kucerakk — APPROVED** [review]

**Review by lego963 — APPROVED** [review]

**Review by gtema — APPROVED** [review]

**Review by irina-pereiaslavskaia — APPROVED** [review]

**Pull request merged**